c_models/makefiles/extra.mk

```diff
@@ -0,0 +1,148 @@
+# This function, if defined, is called from within convert_to_modified.
+#
+define extra_convert_to_modified
+$(patsubst $(EXTRA_SRC_DIR)%,$(EXTRA_MODIFIED_DIR)%,$(1))
+endef
+
+# ----------------------------------------------------------------------
+# Support function for converting modified C paths to object paths.
+#
+# $(NEURAL_MODELLING_DIRS)..... is done by neural_build.mk so not needed here.
+# This function, if defined, is called from within convert_to_object.
+#
+define extra_convert_to_object
+$(patsubst $(EXTRA_MODIFIED_DIR)%.c, $(BUILD_DIR)%.o, $(1))
+endef
+
+# ----------------------------------------------------------------------
+
+include $(NEURAL_MODELLING_DIRS)/makefiles/neuron/neural_build.mk
+
+# ----------------------------------------------------------------------
+
+# EXTRA_SYNAPSE_TYPE_OBJECTS is deprecated!
+# Instead add a new rule
+#$(BUILD_DIR).....o: $(EXTRA_MODIFIED_DIR)...c $(C_FILES_MODIFIED)
+#	-mkdir -p $(dir $@)
+#	$(SYNAPSE_TYPE_COMPILE) -o $@ make$<
+
+                       
+# EXTRA_STDP is deprecated
+# Instead add a new rule
+#$(BUILD_DIR).....o: $(EXTRA_MODIFIED_DIR)...c $(C_FILES_MODIFIED)
+#	-mkdir -p $(dir $@)
+#	$(STDP_COMPILE) -o $@ $<
+
+# ----------------------------------------------------------------------
+# Make rules to copy and transform the extra code (so that logging messages are
+# compressed on SpiNNaker).
+#
+$(EXTRA_MODIFIED_DIR)%.c: $(EXTRA_SRC_DIR)%.c
+	python -m spinn_utilities.make_tools.converter $(EXTRA_SRC_DIR) $(EXTRA_MODIFIED_DIR) $(EXTRA_LOG_DICT_FILE)
+
+$(EXTRA_MODIFIED_DIR)%.h: $(EXTRA_SRC_DIR)%.h
+	python -m spinn_utilities.make_tools.converter $(EXTRA_SRC_DIR) $(EXTRA_MODIFIED_DIR) $(EXTRA_LOG_DICT_FILE)
+
+$(EXTRA_LOG_DICT_FILE): $(EXTRA_SRC_DIR)
+	python -m spinn_utilities.make_tools.converter $(EXTRA_SRC_DIR) $(EXTRA_MODIFIED_DIR) $(EXTRA_LOG_DICT_FILE)
+
+# ----------------------------------------------------------------------
+# Build rule for the transformed extra source files.
+#
+$(BUILD_DIR)%.o: $(EXTRA_MODIFIED_DIR)%.c $(C_FILES_MODIFIED)
+	# extra build
+	-mkdir -p $(dir $@)
+	$(CC) $(CFLAGS) -o $@ $<
+
+# ----------------------------------------------------------------------
+# Add the directories containing the extra transformed sources to the clean
+# rule.
+#
+EXTRA_CLEAN_DIRS += $(EXTRA_MODIFIED_DIR)
+
+# ----------------------------------------------------------------------
+# Optional: Add these to the precious rule so they are not automatically
+# removed by make.
+#
+EXTRA_PRECIOUS += $(EXTRA_MODIFIED_DIR)%.c $(EXTRA_MODIFIED_DIR)%.h $(EXTRA_LOG_DICT_FILE)
+
+# NOTE:
+# $(NEURAL_MODELLING_DIRS)/makefiles/neuron/neural_build.mk brings in all code
+# for $(NEURAL_MODELLING_DIRS).....
+
+# ----------------------------------------------------------------------
+# If more that one extra source directory (not children) is required here, you
+# need to do the following for each source.
+#
+#  1. Define a directory to place the modified code into (like
```

Comment:
why are these steps not automatic? seems a lot of work and a lot of irrleveant stuff for end suers to understand.

Reply:
This is a balance between readable makefiles and automatic ones which are very hard to read.

To do this automatically you would have to have functions which define functions

This can be done later if there is a demand.